src/lib/vasi-sync/src/atomic_tls_map.rs

```diff
@@ -217,6 +221,29 @@ where
         self.keys[idx].store(None, atomic::Ordering::Release);
         Some(value)
     }
+
+    /// # Safety
+    ///
+    /// There must be no outstanding references from `self`, nor parallel access
+    /// to `self` before this method returns.
+    pub unsafe fn clear(&self) {
+        for idx in 0..N {
+            if self.keys[idx]
+                .swap(None, atomic::Ordering::Relaxed)
+                .is_some()
+            {
+                self.values[idx].get_mut().with(|value| {
+                    assert_eq!(self.refcounts[idx].get(), 0);
+                    // SAFETY: Caller has guaranteed that we effectively have exclusive access
+                    // to `self`.
+                    let value = unsafe { &mut *value };
+                    // SAFETY: The `some` value for the key indicates that this value
+                    // is initialized.
+                    unsafe { value.assume_init_drop() }
```

Comment:
The `AtomicTlsMap` doesn't provide any synchronization for writes since it's assumed that an entry is only accessed by a single thread. So I don't see how we can drop the values here in a different thread without synchronization.

Reply:
Good catch.

I have an alternate idea; will probably make a new PR